File: 07-trees/230-kth-smallest-element-in-a-bst/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void inOrderTraversal(TreeNode* root, int& nodeNum, int& result) {
        if(!root || nodeNum == 0)   { //either node is null, or we already hit target
            return;
        }
        inOrderTraversal(root->left, nodeNum, result);
        //we check are we the Kth node, if not
        nodeNum--; 
        if(nodeNum == 0)    {
            result = root->val;
            return;
        }

        inOrderTraversal(root->right, nodeNum, result);
    }
    int kthSmallest(TreeNode* root, int k) {
        /*
        Do an in order traversal, LNR, and 
        the kth node we visit is the kth smallest
        */
        int result = -1;
        int nodeNum = k;
        inOrderTraversal(root, nodeNum, result);
        return result;


    }
};

File: 07-trees/230-kth-smallest-element-in-a-bst/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        #in order traversal, and store each node value in an array
        # index the kth element
        temp_arr = []
        def recursive_func(root):
            #LNR for InOrder Traversal
            if root is None:
                return
            recursive_func(root.left)
            temp_arr.append(root.val)
            recursive_func(root.right)
        
        recursive_func(root)
        return temp_arr[k-1]
            
